homeassistant/components/history.py

```diff
@@ -119,19 +119,39 @@ def get_states(hass, utc_point_in_time, entity_ids=None, run=None,
     from sqlalchemy import and_, func
 
     with session_scope(hass=hass) as session:
-        most_recent_state_ids = session.query(
-            func.max(States.state_id).label('max_state_id')
-        ).filter(
-            (States.created >= run.start) &
-            (States.created < utc_point_in_time) &
-            (~States.domain.in_(IGNORE_DOMAINS)))
+        if entity_ids and len(entity_ids) == 1:
+            # Use an entirely different (and extremely fast) query if we only
+            # have a single entity id
+            most_recent_state_ids = session.query(
+                States.state_id.label('max_state_id')
+            ).filter(
+                (States.created < utc_point_in_time) &
+                (States.entity_id.in_(entity_ids))
+            ).order_by(
+                States.created.desc())
+
+            if filters:
+                most_recent_state_ids = filters.apply(most_recent_state_ids,
+                                                      entity_ids)
+
+            most_recent_state_ids = most_recent_state_ids.limit(1)
 
-        if filters:
-            most_recent_state_ids = filters.apply(most_recent_state_ids,
-                                                  entity_ids)
+        else:
+            # We have more than one entity to look at, so we need to do a much
+            most_recent_state_ids = session.query(
+                func.max(States.state_id).label('max_state_id')
+            ).filter(
+                (States.created >= run.start) &
+                (States.created < utc_point_in_time))
```

Comment:
That was actually intentional (I put it in its own commit to call it out) as it was causing the domain filter to appear twice in every query, due to the fact that the `filters.apply` process applies the IGNORE_DOMAINS itself. This was the only method that had this double-logic so I figured it was safe to remove it and cleans things up a tad.

Interestingly (and this is a topic for a refactor), it appears that the `filters.apply` logic *always* happens when these methods are used in production. The unit tests are the only code that skips the scaffolding and goes straight to the methods without supplying a `filters` object.